compiler/rustc_mir/src/interpret/step.rs

```diff
@@ -113,6 +113,35 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 M::retag(self, *kind, dest)?;
             }
 
+            // Call CopyNonOverlapping
+            CopyNonOverlapping(box rustc_middle::mir::CopyNonOverlapping { dst, src, count }) => {
+                let count = self.eval_operand(count, None)?;
+                let count = self.read_scalar(count)?.to_machine_usize(self)?;
+
+                let src = self.eval_operand(src, None)?;
+                let dst = self.eval_operand(dst, None)?;
+                let layout =
+                    src.layout.pointee_info_at(self, rustc_target::abi::Size::ZERO).unwrap();
+                let (size, align) = (layout.size, layout.align);
+                if count != 0 && size.bytes() != 0 {
```

Comment:
This is incorrect. It will mean that dangling pointers (i.e., pointers to memory that was previously allocated but has been deallocated) will not be detected properly for zero-sized copies.

See https://github.com/rust-lang/rust/blob/090dac0c3892a2d21afb0f82aa72920760e08078/compiler/rustc_mir/src/interpret/intrinsics.rs#L326 for a correct implementation of `copy_nonoverlapping`. Is it possible to share that code by moving it to a helper function?

Reply:
@nagisa FWIW, I think it'd be good to Cc @rust-lang/wg-const-eval on such fundamental changes to the Miri/CTFE engine like adding support for a new statement or intrinsic. I would almost have missed this PR.

Reply:
Ah so I should just get rid of this if branch entirely because an error should be propagated in certain cases?

Reply:
Here we don't have access to substs, and only keep the Operand from OpTy

Reply:
Right, so getting the `elem_layout` works a bit differently. Your approach for that is fine.

So you should be able to write a function like
```rust
fn copy_nonoverlapping(&mut self, src: mir::OpTy, dst: mir::OpTy, count: mir::OpTy) { ... }
```
and call that from both places, right? It should work mostly like the intrinsic currently does, except using your approach to get the `elem_layout`.

Reply:
I don't think `copy_nonoverlapping` is ever called from `compiler/rustc_mir/src/interpret/intrinsics.rs` now, so is there a point to moving this to a function to be called in both places?

Reply:
Why is it not called though, wasn't the entire point to share that code? I am confused.

Also... is that intrinsic code even still reachable? I did not realize that you are entirely lowering the intrinsic away in `lower_intrinsics.rs`, so it might be that we can just remove the intrinsic implementation (in all backends -- Miri and the codegen backends).

Reply:
The original point of this PR was to write a MIR statement which is more amenable to optimization because it's known to terminate. I can remove some implementations if it makes sense, but to be honest I'm not fully sure where everything is being used so I just added them all initially.